docs/technical/git-submodules.md

```diff
@@ -0,0 +1,156 @@
+
+  - `submodule<submodule-name>.url` contains the url to clone the submodule
+    from.
+
+In the working tree, Git notices the presence of a submodule by the `.git` entry
+(signifying the root of a Git repository working tree). This is either the
+submodule's actual Git directory (an "old-form" submodule), or a `.git` file
+pointing to `<superproject-git-directory>/modules/<submodule-name>`. The latter
+is sometimes called the "absorbed form", and is Git's preferred mode of
+operation.
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+Colocated repositories will be supported where they are convenient, but they are
+not a high priority.
+
+This section should be treated as a set of guidelines, not a strict order of
+work.
+
+### Phase 1: Readonly submodules
+
+This includes work that inspects submodule contents but does not create new
+objects in the submodule. This requires a way to store submodules in a jj
+repository that supports readonly operations.
+
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
+
+### Phase 2: Snapshotting new changes
+
+This allows a user to write new contents to a submodule and its remote.
+
+#### Outcomes
+
+- Changes in the working copy can be recorded in a submodule commit
+- Submodule branches can be modified
+- Submodules and their branches can be pushed to their remote
+
+### Phase 3: Merging/rebasing/conflicts
+
+This allows merging and rebasing of superproject commits in a content-aware way
+(in contrast to Git, where only the gitlink commit ids are compared), as well as
+workflows that make resolving conflicts easy and sensible.
+
+This can be done in tandem with Phase 2, but will likely require a significant
+amount of design work on its own.
+
+#### Outcomes
+
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: Possible, but unprioritized work
+
+- Rewriting submodule commits and updating superproject gitlinks accordingly
+
+## Design
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
```

Comment:
Can we consider other backends before talking about the Git backend? In previous discussions it was always a hot point as we always respected Google's internal backend.

Reply:
I think I will also specify non-Git backends as a non-goal.

Though I'm not working on Google's internal backend, I feel confident saying that it will ~never need to support Git submodules.

Other, future backends are probably too ill-defined to be worth considering (e.g. Would these other backends have Git submodules? How would they convert Git submodules into their native format?). We could reconsider the feature if we start supporting non-Git 'submodules', but perhaps the general form would look very different from this proposal.

Reply:
> I think I will also specify non-Git backends as a non-goal.

Thats fair but I'd still like to discuss the impact of your changes to the base backend. This is to avoid leaking internals to other backends and building a common understanding, of what a backend is/implies. The goal here is to take a step back and minimize the impact on the base backend.

> Though I'm not working on Google's internal backend, I feel confident saying that it will ~never need to support Git submodules.

I also don't expect something Perforce-like to ever grow a understanding of Git submodules. But Git submodules cannot constrain it, or any other backend. 

> Other, future backends are probably too ill-defined to be worth considering (e.g. Would these other backends have Git submodules? How would they convert Git submodules into their native format?). We could reconsider the feature if we start supporting non-Git 'submodules', but perhaps the general form would look very different from this proposal.

I don't expect that any other backend will implement something like Git submodules or Jujutsu submodules if they ever exist. So it's not immediately worth discussing, but we'd still learn something while doing some space exploration. But its totally out of scope for this proposal.

Reply:
> Thats fair but I'd still like to discuss the impact of your changes to the base backend. This is to avoid leaking internals to other backends and building a common understanding, of what a backend is/implies. The goal here is to take a step back and minimize the impact on the base backend.

Yes, I've considered that as well, though I'm not concerned since there are design patterns that can manage this. A successful one IMO is "capabilities", where we could specify a set of capabilities that a backend may or may not support, and Git submodules can be one of them. Other backends that want to support Git submodules can provide their own implementation, but it won't constrain backends that don't support submodules.

But this does raise the question of whether the capability should be "Git submodules", or a generalized form ("Subrepos"?). I'm leaning towards "Git submodules", because I think it would be difficult to generalize "subrepos" across all possible backends.

That said, there is an alternative in this space..

> I don't expect that any other backend will implement something like Git submodules or Jujutsu submodules if they ever exist. So it's not immediately worth discussing, but we'd still learn something while doing some space exploration. But its totally out of scope for this proposal.

Which is to implement Git submodules as jj repositories, as @hanwen suggested https://github.com/martinvonz/jj/pull/1411#discussion_r1144994721. I think that already gets us to generalized "subrepos", because we can tell the subrepo what to do and not worry about its backend.

My biggest concern with that proposal is that the layout on disk would probably be different from what Git expects, so it might make colocated repos harder to implement. _But_ I feel fairly confident that we can upstream changes Git and libgit2 to understand the different layout, so I'm not that worried.

Reply:
> Yes, I've considered that as well, though I'm not concerned since there are design patterns that can manage this. A successful one IMO is "capabilities", where we could specify a set of capabilities that a backend may or may not support, and Git submodules can be one of them. Other backends that want to support Git submodules can provide their own implementation, but it won't constrain backends that don't support submodules.

I'm not directly concerned with the trait or interface but rather with the concept. I agree that there are many design patterns which help to solve this issue.

> But this does raise the question of whether the capability should be "Git submodules", or a generalized form ("Subrepos"?). I'm leaning towards "Git submodules", because I think it would be difficult to generalize "subrepos" across all possible backends.
[...]
> Which is to implement Git submodules as jj repositories, as @hanwen suggested https://github.com/martinvonz/jj/pull/1411#discussion_r1144994721. I think that already gets us to generalized "subrepos", because we can tell the subrepo what to do and not worry about its backend.

This sounds like something for `Alternatives Considered` section, if we were discussing a JJ submodules proposal, its also very neat. 

> My biggest concern with that proposal is that the layout on disk would probably be different from what Git expects, so it might make colocated repos harder to implement. But I feel fairly confident that we can upstream changes Git and libgit2 to understand the different layout, so I'm not that worried.

I'm also not worried about a disk format yet, as this is something for the detailed implementation plan. I assume your capable of doing the necessary changes in Git and JJ, since you're a Git contributor.

Reply:
> I'm not directly concerned with the trait or interface but rather with the concept.

Hm, I suppose I'm having trouble imagining what you're describing here. Could you perhaps give an example sentence or two of what you wish I had included?

At any rate, I'm thinking of taking a step back from the idea of having Git submodules in the Git backend only. I suggested it not because it follows from the primary requirements, but because I thought it was uncontroversial for submodules to not be jj repos (based on informal discussions), so this would be the easiest way to get to colocated support . However, @hanwen has given some good reasons to think otherwise, so I'll probably revise this.

Reply:
Sorry for being so obtuse, I was a bit caught in my own thoughts. 

>> I'm not directly concerned with the trait or interface but rather with the concept.

>Hm, I suppose I'm having trouble imagining what you're describing here. Could you perhaps give an example sentence or two of what you wish I had included?

In my opinion, you went straight to a implementation with your idea (which is fine), but you brushed my concerns about a larger impact on Jujutsu aside. This by moving it out of scope by making it Git only[^1] or capabilites[^2].  

I just wanted to discuss the benefits or drawbacks for Jujutsu and other backends, without concerns about how an implementation would look 

> At any rate, I'm thinking of taking a step back from the idea of having Git submodules in the Git backend only. I suggested it not because it follows from the primary requirements, but because I thought it was uncontroversial for submodules to not be jj repos (based on informal discussions), so this would be the easiest way to get to colocated support . However, @hanwen has given some good reasons to think otherwise, so I'll probably revise this.  

I also think @hanwen found a good solution and I'll be happy to discuss it further after an Google internal discussion.

> so, in general, your doc starts off with a set of requirements, which I think is good, but in this section you're already pinning yourself to a technical decision which will have a lot of future implications. Isn't it too early to make a call here? Ideally, the decision follows from requirements.

This also reflects my opinion. 

Sorry for the Wall of Text, I just wanted to leave no room for misinterpretations. 


[^1]: Aside, then the proposal should be called `Git submodules for the Git backend`, as the goal is to improve the UX.
[^2]: Which address the problem on an interface/trait level, sidestepping the discussion on potential global impact.

Reply:
> I'm sorry if I gave that impression. That was absolutely not my intention; it's quite clear now that I misinterpreted you quite a lot.

No harm done :-), we just talked past each other.

 > I'm afraid I still don't completely follow :/ I read this comment as "the benefit or drawbacks for Jujutsu and other backends (of supporting Git submodules), but rereading your initial comment, my impression is that you were talking about what considering support [Git] submodules in other backends.

You got the point, which is important. Even writing two to three sentences about other backends and dismissing them would've been enough, to bring out concerns which weren't considered before. 

> Tbh I'd prefer for the discussion to not be internal-only. After all, we aren't the only ones using jj :)

I just deferred to the internal discussion as you have some Git submodule users already (Android, Gerrit and Chromium in a year or two). Its helpful for everyone if you'd evaluate some proposed solutions against your use cases. 
As I'm neither user nor have any direct use case for Git submodules.   

> If there are no objections, I'll scale this PR back to the truly uncontroversial bits (which seem to be requirements + roadmap). Then, I'll focus my next PR on figuring out the criteria that we'll use to decide which solution to go with, and evaluating possible approaches.

I do not have any further objections. 

Although an experience report of your Submodule work in git/git could be useful in the next PR, but that's totally up to you.

Reply:
> Even writing two to three sentences about other backends and dismissing them would've been enough, to bring out concerns which weren't considered before.

Ah, yes, that makes sense. I was counting on the shared context that others + myself built up over informal conversations, but now I see how jarring this is if you're not privy to that context.

> I do not have any further objections.

Thanks. Let's move this discussion to the follow up PR :)